topology/nhlt/intel/ssp-nhlt.c

```diff
@@ -0,0 +1,306 @@
+		f_conf1[i].format.nBlockAlign = channel_count * bits_per_sample / 8;
+		f_conf1[i].format.nAvgBytesPerSec = sample_rate * f_conf1[i].format.nBlockAlign;
+
+		/* bytes after this value in this struct */
+		f_conf1[i].format.cbSize = 22;
+		/* actual bits in container */
+		f_conf1[i].format.wValidBitsPerSample = 0;
+		/* channel map not used at this time */
+		f_conf1[i].format.dwChannelMask = 0;
+		/* WAVE_FORMAT_PCM guid (0x0001) ? */
+		f_conf1[i].format.SubFormat[0] = 0;
+		f_conf1[i].format.SubFormat[1] = 0;
+		f_conf1[i].format.SubFormat[2] = 0;
+		f_conf1[i].format.SubFormat[3] = 0;
+
+		ret = ssp_get_vendor_blob_size(nhlt, &blob_size);
+		if (ret < 0) {
+			fprintf(stderr, "nhlt_ssp_get_ep: dmic_get_vendor_blob_size failed\n");
+			return ret;
+		}
+		f_conf1[i].vendor_blob.capabilities_size = blob_size;
+	}
+
+	ep.length = sizeof(struct endpoint_descriptor) +
+		sizeof(struct ssp_device_specific_config) +
+		sizeof(struct formats_config) +
+		sizeof(struct format_config) * f_conf.formats_count +
+		blob_size * f_conf.formats_count;
+
+	/* allocate the final variable length ep struct */
+	ep_target = calloc(ep.length, sizeof(uint8_t));
+	if (!ep_target)
+		return -ENOMEM;
+
+	*eps = (struct endpoint_descriptor *)ep_target;
+
+	/* copy all parsed sub arrays into the top level array */
+	memcpy(ep_target, &ep, sizeof(struct endpoint_descriptor));
+
+	ep_target += sizeof(struct endpoint_descriptor);
+
+	memcpy(ep_target, &ssp_conf, sizeof(struct ssp_device_specific_config));
+	ep_target += sizeof(struct ssp_device_specific_config);
+
+	memcpy(ep_target, &f_conf, sizeof(struct formats_config));
+	ep_target += sizeof(struct formats_config);
+
+	/* copy all hw configs */
+	for (i = 0; i < f_conf.formats_count; i++) {
+		memcpy(ep_target, &f_conf1[i], sizeof(struct format_config));
+		ep_target += sizeof(struct format_config);
+		ret = ssp_get_vendor_blob(nhlt, ep_target, dai_index, i);
+		if (ret < 0) {
+			fprintf(stderr, "nhlt_sso_get_ep: ssp_get_vendor_blob failed\n");
+			return ret;
+		}
+		ep_target += blob_size;
+	}
+
+	return 0;
+}
+
+/* Set ssp parameters from topology for ssp coefficient calculation.
+ *
+ * You can see an example of topology v2 config of ssp below. In this example the default
+ * object parameters are spelled out for clarity. General parameters like sample_bits are parsed
+ * with set_ssp_data and hw_config object data with set_hw_data. Ssp can have multiple hw_configs.
+ * Values are saved into intermediate structs and the vendor specific blob is calculated at the end
+ * of parsing with ssp_calculate.
+ *
+ * 	SSP."0" {
+ *		id 			0
+ *		direction		"duplex"
+ *		name			NoCodec-0
+ *		default_hw_conf_id	0
+ *		sample_bits		16
+ *		quirks			"lbm_mode"
+ *		bclk_delay		0
+ *		mclk_id 		0
+ *		default_hw_config_id	0
+ *		clks_control 		0
+ *		frame_pulse_width	0
+ *		tdm_padding_per_slot	false
+ *
+ *		Object.Base.hw_config."SSP0" {
+ *			id	0
+ *			mclk_freq	38400000
+ *			bclk_freq	4800000
+ *			tdm_slot_width	32
```

Comment:
this example cannot work, the bclk would be 3072000 Hz...

Reply:
oh, maybe bclk 2400000

Reply:
nope, blck of 2.4 MHz cannot work because that would mean slots of 25 bits that NHLT cannot represent.

for NHLT, you have to assume the slot_width is a multiple of 8, and then find the bclk=fsync * number of slots * slots_width.

that will not work of course for platforms with a 24 MHz root clock, where I think the only solution might be to use the dreaded M/N dividers.